protocol/communities/manager.go

```diff
@@ -1433,25 +1511,91 @@ func (m *Manager) checkPermissionToJoin(permissions []*protobuf.CommunityTokenPe
-	hasERC20 := false
-	for _, tokenRequirement := range tokenCriteria {
-		// we gotta check for whether there are ERC20 token criteria
-		// in the first place, if we don't we'll return a false positive
-		if tokenRequirement.Type == protobuf.CommunityTokenType_ERC20 {
-			hasERC20 = true
-			amount, _ := strconv.ParseFloat(tokenRequirement.Amount, 32)
-			if balances[tokenRequirement.Symbol].Cmp(big.NewFloat(amount)) == -1 {
-				result = false
-				break
+func extractTokenRequirements(permissions []*protobuf.CommunityTokenPermission) (map[uint64]map[string]*protobuf.TokenCriteria, map[uint64]map[string]*protobuf.TokenCriteria) {
+	erc20TokenRequirementsByChain := make(map[uint64]map[string]*protobuf.TokenCriteria)
+	erc721TokenRequirementsByChain := make(map[uint64]map[string]*protobuf.TokenCriteria)
+	for _, tokenPermission := range permissions {
+		for _, tokenRequirement := range tokenPermission.TokenCriteria {
+			isERC721 := tokenRequirement.Type == protobuf.CommunityTokenType_ERC721
+			isERC20 := tokenRequirement.Type == protobuf.CommunityTokenType_ERC20
+			for chainID, contractAddress := range tokenRequirement.ContractAddresses {
+
+				_, existsERC721 := erc721TokenRequirementsByChain[chainID]
+
+				if isERC721 && !existsERC721 {
+					erc721TokenRequirementsByChain[chainID] = make(map[string]*protobuf.TokenCriteria)
+				}
+				_, existsERC20 := erc20TokenRequirementsByChain[chainID]
+
+				if isERC20 && !existsERC20 {
+					erc20TokenRequirementsByChain[chainID] = make(map[string]*protobuf.TokenCriteria)
+				}
+
+				_, existsERC721 = erc721TokenRequirementsByChain[chainID][contractAddress]
+				if isERC721 && !existsERC721 {
+					erc721TokenRequirementsByChain[chainID][strings.ToLower(contractAddress)] = tokenRequirement
+				}
+
+				_, existsERC20 = erc20TokenRequirementsByChain[chainID][contractAddress]
+				if isERC20 && !existsERC20 {
+					erc20TokenRequirementsByChain[chainID][strings.ToLower(contractAddress)] = tokenRequirement
+				}
+			}
+		}
+	}
+	return erc20TokenRequirementsByChain, erc721TokenRequirementsByChain
+}
+
+func (m *Manager) getOwnedERC721Tokens(walletAddresses []gethcommon.Address, tokenRequirements map[uint64]map[string]*protobuf.TokenCriteria) (map[uint64]map[string][]opensea.Asset, error) {
+
+	if m.walletConfig == nil || len(m.walletConfig.OpenseaAPIKey) == 0 {
+		return nil, errors.New("no api key for opensea")
+	}
+
+	ownedERC721Tokens := make(map[uint64]map[string][]opensea.Asset)
+
+	for chainID, erc721Tokens := range tokenRequirements {
+		contractAddresses := make([]gethcommon.Address, 0)
+		for contractAddress := range erc721Tokens {
+			contractAddresses = append(contractAddresses, gethcommon.HexToAddress(contractAddress))
+		}
+
+		if _, exists := ownedERC721Tokens[chainID]; !exists {
+			ownedERC721Tokens[chainID] = make(map[string][]opensea.Asset)
+		}
+
+		for _, owner := range walletAddresses {
+			client, err := opensea.NewOpenseaClient(chainID, m.walletConfig.OpenseaAPIKey)
```

Comment:
Do we need a new client for each wallet address?

Reply:
Yea this has been a bit of a back and forth because of the request limit error we got from opensea, but eventually this has been changed such that we'll always get the same client instance per chain ID, no matter how often you call `NewOpenseaClient`.

In other words, it's kinda like a singleton at this point so it doesn't make a difference. But I can change it since it (rightfully) causes questions.